// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package pcit.public_key_service;

import "google/protobuf/empty.proto";

// Service to fetch public keys.
// Note that another job will be reading from this and publishing the result in
// a JSON file to Cloud CDN. Therefore, the performance or availability of this
// service is not critical.
service PublicKeyService {
  // Fetches latest public keys from KeyService.
  rpc ListPublicKeys(google.protobuf.Empty) returns (ListPublicKeysResponse) {}
  // RPC method to update the cloud bucket containing the public keys.
  rpc UpdateCloudBucket(google.protobuf.Empty) returns (google.protobuf.Empty) {
  }
}

// Response of listing public keys
message ListPublicKeysResponse {
  repeated PublicKey public_keys = 1 [json_name = "keys"];
}

enum ExecutionEnvironment {
  EXECUTION_ENVIRONMENT_UNSPECIFIED = 0;
  EXECUTION_ENVIRONMENT_ONPREM = 1;
}

message PublicKey {
  // The id of the public key.
  string key_id = 1 [json_name = "id"];
  // The value of the public key formated in unsafe web base64.
  string public_key = 2 [json_name = "key"];
  // The identifier for per key origin.
  string origin = 3 [json_name = "origin"];
  // The environment the service runs in.
  ExecutionEnvironment execution_environment = 4
      [json_name = "execution_environment"];
}
